modules/services/heartbeat_monitor.lua

```diff
@@ -3,69 +3,79 @@ local constants = require('protocol_handler/ford_protocol_constants')
+    if self.heartbeatEnabled then
+        if msg.frameType == constants.FRAME_TYPE.CONTROL_FRAME and
+           msg.frameInfo == constants.FRAME_INFO.HEARTBEAT_ACK and
+           self.IgnoreHeartBeatAck.get() then
+            return
         end
-      end
+        self.heartbeatFromSDLTimer:reset()
     end
+  end)
+end
 
-    function d.CloseSession()
-      if self.heartbeatEnabled then
-        print("\27[31m SDL didn't send anything for " .. self.heartbeatFromSDLTimer:interval()
-          .. " msecs. Closing session # " .. self.session.sessionId.get().."\27[0m")
-        self.control_services:StopService(7)
-        :Do(self:StopHeartbeat())
-      end
-    end
+function mt.__index:SendHeartbeatAck()
+  self.control_services:SendControlMessage( {frameInfo = constants.FRAME_INFO.HEARTBEAT_ACK } )
+end
 
-    self.session.connection:OnInputData(function(_, msg)
-      if self.session.sessionId.get() ~= msg.sessionId then return end
-      if self.heartbeatEnabled then
-          if msg.frameType == constants.FRAME_TYPE.CONTROL_FRAME and
-             msg.frameInfo == constants.FRAME_INFO.HEARTBEAT_ACK and
-             self.IgnoreHeartBeatAck.get() then
-              return
-          end
-          self.heartbeatFromSDLTimer:reset()
+function mt.__index:AddHeartbeatExpectation()
+  local event = events.Event()
+  event.matches = function(s, data)
+    return data.frameType == constants.FRAME_TYPE.CONTROL_FRAME and
+    data.serviceType == constants.SERVICE_TYPE.CONTROL and
+    data.frameInfo == constants.FRAME_INFO.HEARTBEAT   and
+    self.session.sessionId.get() == data.sessionId
+  end
+  self.expectations:ExpectEvent(event, "Heartbeat")
+  :Pin()
+  :Times(AnyNumber())
+  :Do(function(data)
+      if self.heartbeatEnabled and self.AnswerHeartbeatFromSDL.get() then
+        self:SendHeartbeatAck()
       end
     end)
-
-    qt.connect(self.heartbeatToSDLTimer, "timeout()", d, "SendHeartbeat()")
-    qt.connect(self.heartbeatFromSDLTimer, "timeout()", d, "CloseSession()")
 end
 
+
 function mt.__index:StartHeartbeat()
-  if self.heartbeatToSDLTimer and self.heartbeatFromSDLTimer then
-    self.heartbeatEnabled = true
+  self.heartbeatEnabled = true
+
+  function d.SendHeartbeat()
+    if self.heartbeatEnabled and self.SendHeartbeatToSDL.get() then
+      self.control_services:SendControlMessage( { frameInfo = constants.FRAME_INFO.HEARTBEAT } )
+      if not self.IgnoreHeartBeatAck.get() then
+        self.heartbeatToSDLTimer:reset()
+      end
+    end
+  end
+
+  function d.CloseSession()
+    if self.heartbeatEnabled then
+      print("\27[31m SDL didn't send anything for " .. self.heartbeatFromSDLTimer:interval()
+        .. " msecs. Closing session # " .. self.session.sessionId.get().."\27[0m")
+      self.control_services:StopService(7)
+      :Do(self:StopHeartbeat())
+    end
+  end
+
+  xmlReporter.AddMessage("StartHearbeat", "True", (config.heartbeatTimeout + 1000))
```

Comment:
@VVeremjova please extract config.heartbeatTimeout + 1000 to variable to avoid repeating magic numbers